pkg/devfile/parser/data/v2/attributes_test.go

```diff
@@ -0,0 +1,197 @@
+				}
+			}
+		})
+	}
+}
+
+func TestUpdateAttributes(t *testing.T) {
+
+	tests := []struct {
+		name       string
+		devfilev2  *DevfileV2
+		attributes attributes.Attributes
+		wantErr    bool
+	}{
+		{
+			name: "Schema 2.0.0 does not have attributes",
+			devfilev2: &DevfileV2{
+				v1alpha2.Devfile{
+					DevfileHeader: devfilepkg.DevfileHeader{
+						SchemaVersion: "2.0.0",
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "Schema 2.1.0 has attributes",
+			devfilev2: &DevfileV2{
+				v1alpha2.Devfile{
+					DevfileHeader: devfilepkg.DevfileHeader{
+						SchemaVersion: "2.1.0",
+					},
+					DevWorkspaceTemplateSpec: v1alpha2.DevWorkspaceTemplateSpec{
+						DevWorkspaceTemplateSpecContent: v1alpha2.DevWorkspaceTemplateSpecContent{
+							Attributes: attributes.Attributes{}.PutString("key1", "value1").PutString("key2", "value2"),
+						},
+					},
+				},
+			},
+			attributes: attributes.Attributes{}.PutString("key3", "value3"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.devfilev2.UpdateAttributes(tt.attributes)
+			if tt.wantErr == (err == nil) {
+				t.Errorf("TestUpdateAttributes error - %v, wantErr %v", err, tt.wantErr)
+			} else if err == nil {
+				attributes, err := tt.devfilev2.GetAttributes()
+				if err != nil {
+					t.Errorf("TestUpdateAttributes error2 - %+v", err)
+				}
+				if !reflect.DeepEqual(attributes, tt.attributes) {
+					t.Errorf("TestUpdateAttributes mismatch error - expected %+v, actual %+v", tt.attributes, attributes)
+				}
+			}
+		})
+	}
+}
+
+func TestAddAttributes(t *testing.T) {
+
+	nestedValue := map[string]interface{}{
+		"key3.1": map[string]interface{}{
+			"key3.1.1": "value3.1.1",
+		},
+	}
+
+	tests := []struct {
+		name           string
+		devfilev2      *DevfileV2
+		key            string
+		value          interface{}
+		wantAttributes attributes.Attributes
+		wantErr        bool
+	}{
+		{
+			name: "Schema 2.0.0 does not have attributes",
+			devfilev2: &DevfileV2{
+				v1alpha2.Devfile{
+					DevfileHeader: devfilepkg.DevfileHeader{
+						SchemaVersion: "2.0.0",
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "Schema 2.1.0 has attributes",
+			devfilev2: &DevfileV2{
+				v1alpha2.Devfile{
+					DevfileHeader: devfilepkg.DevfileHeader{
+						SchemaVersion: "2.1.0",
+					},
+					DevWorkspaceTemplateSpec: v1alpha2.DevWorkspaceTemplateSpec{
+						DevWorkspaceTemplateSpecContent: v1alpha2.DevWorkspaceTemplateSpecContent{
+							Attributes: attributes.Attributes{}.PutString("key1", "value1").PutString("key2", "value2"),
+						},
+					},
+				},
+			},
+			key:            "key3",
+			value:          nestedValue,
+			wantAttributes: attributes.Attributes{}.PutString("key1", "value1").Put("key3", nestedValue, nil).PutString("key2", "value2"),
+		},
+		{
+			name: "If Schema 2.1.0 has an attribute already present, it should overwrite",
```

Comment:
I'm not sure if this is an expected behavior for `add`
if you check for other update & add functions for commands/components/projects etc., this should be an update behavior. for "add", if the key exist, should return an error?

Reply:
Do these modifications:
update will look for a specific key, err out if absent
add will just add a new key, since attributes is a map if there is a key already present, it will just overwrite with the new value